ops/framework.py

```diff
@@ -254,15 +294,27 @@ class SomeObject(Object):
     attribute which is a BoundEvent and may be used to emit and observe the event.
     """
 
-    def __init__(self, event_type):
+    if TYPE_CHECKING:
+        # to help the type checker and IDEs:
+        # these attributes are only present on the bound descriptor object
+        @property
+        def event_kind(self) -> str: ...  # noqa
+        @event_kind.setter
+        def event_kind(self, val: str) -> None: ...  # noqa
+        @property
+        def emitter_type(self) -> 'Type[Object]': ...  # noqa
+        @emitter_type.setter
+        def emitter_type(self, val: 'Type[Object]') -> None: ...  # noqa
+
+    def __init__(self, event_type: 'Type[_EventType]'):
         if not isinstance(event_type, type) or not issubclass(event_type, EventBase):
             raise RuntimeError(
                 'Event requires a subclass of EventBase as an argument, got {}'.format(event_type))
         self.event_type = event_type
         self.event_kind = None
         self.emitter_type = None
```

Comment:
It doesn't require Optional because it's a class instance (not sure WHY that is in turn though), if you move those to __init__, it does complain (not pyright; but pycharm at least)